File: services/work-failer/app/util/env.ts
import { IsInt, Min } from 'class-validator';
import * as path from 'path';

import { HarmonyEnv } from '@harmony/util/env';

//
// env module
// Sets up the environment variables for the work failer using the base environment variables
// and some specific to the work failer
//

class FailerHarmonyEnv extends HarmonyEnv {

  @IsInt()
  @Min(1)
  workFailerPeriodSec: number;

  @IsInt()
  @Min(1)
  workFailerBatchSize: number;

  @IsInt()
  @Min(1)
  failableWorkAgeMinutes: number;

  @IsInt()
  @Min(-1)
  maxWorkItemsOnUpdateQueueFailer: number;

  @IsInt()
  @Min(1)
  defaultTimeoutSeconds: number;
}

const localPath = path.resolve(__dirname, '../../env-defaults');
const failerHarmonyEnvObj = new FailerHarmonyEnv(localPath);
failerHarmonyEnvObj.validate();

export default failerHarmonyEnvObj;

File: core-services/internal/cronjobs/workfailer/env.go
// Environment variables for the work failer
package workfailer

import (
	"fmt"
	"os"
	"strconv"

	"github.com/go-playground/validator/v10"
	"github.com/nasa/harmony/core-services/internal/env"
)

type WorkFailerEnv struct {
	env.HarmonyEnv
	WorkFailerCron                  string `validate:"required,cron"`
	WorkFailerBatchSize             int    `validate:"gte=1"`
	FailableWorkAgeMinutes          int    `validate:"gte=1"`
	MaxWorkItemsOnUpdateQueueFailer int    `validate:"gte1"`
}

func InitEnv(baseEnv env.HarmonyEnv) WorkFailerEnv {
	cronEntry := os.Getenv("WORK_FAILER_CRON")
	if cronEntry == "" {
		cronEntry = WORK_FAILER_CRON
	}

	batchSize, err := strconv.Atoi(os.Getenv("WORK_FAILER_BATCH_SIZE"))
	if err != nil {
		batchSize = WORK_FAILER_BATCH_SIZE
	}
	failableWorkAgeMinutes, err := strconv.Atoi(os.Getenv("FAILABLE_WORK_AGE_MINUTES"))
	if err != nil {
		failableWorkAgeMinutes = FAILABLE_WORK_AGE_MINUTES
	}
	maxWorkItemsOnUpdateQueueFailer, err := strconv.Atoi(os.Getenv("MAX_WORK_ITEMS_ON_UPDATE_QUEUE_FAILER"))
	if err != nil {
		maxWorkItemsOnUpdateQueueFailer = MAX_WORK_ITEMS_ON_UPDATE_QUEUE_FAILER
	}

	env := WorkFailerEnv{
		baseEnv,
		cronEntry,
		batchSize,
		failableWorkAgeMinutes,
		maxWorkItemsOnUpdateQueueFailer,
	}

	validate := validator.New(validator.WithRequiredStructEnabled())
	err = validate.Struct(env)
	if err != nil {
		fmt.Println("Invalid work failer env vars")
		panic(err)
	}

	return env

}
